Integrated skip verify, gomail and templates

tboerger opened the pull request:
I have switched the mail sending mechanism to gomail to get more power
out of the sending method. Beside that I have added a flag to skip
certificate verification.

On top of that I have integrated the drone templating to make it
possible to overwrite the email templates optionally.

To be more compatible I have created a plaintext message format with the
help of html2text as well.

crhym3 commented:
What are the use cases that `net/smtp` can't handle and `github.com/go-gomail/gomail` does?

tboerger commented:
> What are the scenarios that net/smtp can't handle and github.com/go-gomail/gomail does?

Getting rid of `x509: certificate signed by unknown authority` is much more complicated with `net/smtp` because you have to rewrite `SendMail` on your own.

crhym3 commented:
That's it?

> Note, however, that this is insecure and should not be used in production.

Which I absolutely agree with, also given letsencrypt and ACME spec in general. This should never be used in production.

tboerger commented:
> That's it?
> 
> Note, however, that this is insecure and should not be used in production.
> Which I absolutely agree with, also given letsencrypt and ACME spec in general. This should never be used in production.

I think it's the decision of the user how he's running his mail servers. We already had a request related to exactly that. I have added another piece to the reasons... easier multipart handling. As you are more than familar with golang, maybe you can give me a better implementation with pure `net/smtp`? I would be happy about it.

crhym3 commented:
Multipart? What about `mime/multipart`.

crhym3 commented:
It's just seems this change will give up too much for a single additional option, which is not even recommended for production use, probably covering very small amount of users who run their own smtp servers (which is a bad idea on its own unless you have a dedicated team managing your smtp).

bradrydzewski commented:
> It's just seems this change will give up too much for a single additional option, which is not even recommended for production use, probably covering very small amount of users who run their own smtp servers (which is a bad idea on its own unless you have a dedicated team managing your smtp).

I sort of agree. The beauty of plugins is that if one person wants to skip SSL verification they can fork the project and make the necessary changes. Plugins don't have to do everything for everyone.

tboerger commented:
But it's also the multipart email, not only the skip verify option. And even just writing multipart with std lib will bloat the code much more.

vtolstov commented:
Im not agree,disabling ssl verify can bring ability to prepare and debug
things before issue any valud cert. Also some services does not have valud
certs, and i need to contact them. Why golang std lub have ability to skip
verify?
23 дек. 2015 г. 20:39 пользователь "Brad Rydzewski" <
notifications@github.com> написал:

> It's just seems this change will give up too much for a single additional
> option, which is not even recommended for production use, probably covering
> very small amount of users who run their own smtp servers (which is a bad
> idea on its own unless you have a dedicated team managing your smtp).
>
> I sort of agree. The beauty of plugins is that if one person wants to skip
> SSL verification they can fork the project and make the necessary changes.
> Plugins don't have to do everything for everyone.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/drone-plugins/drone-email/pull/4#issuecomment-166952950>
> .
>

bradrydzewski commented:
@tboerger this is a simple multi-part gist that I've seen used in the past:

```
From: {{ FROM }}
To: {{ TO }}
Subject: {{ SUBJECT }}
MIME-Version: 1.0
Content-Type: multipart/alternative; boundary="e255104e821101849e9aefa6986"

--e255104e821101849e9aefa6986
Content-Type: text/plain; charset=UTF-8

{{ TEXT MESSAGE }}

--e255104e821101849e9aefa6986
Content-Type: text/html; charset=UTF-8

{{ HTML MESSAGE }}

--e255104e821101849e9aefa6986--
```

bradrydzewski commented:
> Im not agree, disabling ssl verify can bring ability to prepare and debug things before issue any valud cert.

While I agree in principal, does this really apply to Drone? I would expect someone to setup and debug smtp certificates using a mail client or some command line tools, not using Drone.

> Also some services does not have valud certs, and i need to contact them. Why golang std lub have ability to skip verify?

Do you need to contact these services via Drone? I would expect Company X uses Drone and connects to a single Company X smtp server, rather than multiple insecure smtp servers.

While I'm sure there are situations where someone wants to skip TLS verification when sending an email from Drone, I would imagine this is the exception and not the norm. And if it is the exception, someone can always create their own custom plugin.

jackspirou commented:
> Plugins don't have to do everything for everyone ... just most things for most people :smile:

@bradrydzewski I totally agree in principal, but (and being selfish here) I personally would like to see the skip TLS verification that @tboerger is providing. 

At my work, its not so much the mail server that is the issue most of the time, or how they have TLS configured... usually its the damn proxies (and transparent proxies) in between.  I think it is a common pattern to have multiple SMTP servers running behind a loadbalancer/proxy with a DNS name internally.  Not all internal proxies (at my work) have the latest versions/ciphers for TLS, so they often break (man-in-the-middle) or corrupt the request even if you specify options like  `HTTPS_PROXY`.  Having a TLS skip option helps with debugging so that we can tell the folks that run the proxy or other services to upgrade or whitelist that URL/service.  

I would assume other corporations have similar issues maintaining current TLS ciphers and certs, but maybe we are "special" in that sense :stuck_out_tongue_winking_eye: 

In summary I think we can take it or leave it and maybe SMTP should be considered differently, but as a debug option it is nice to have.

bradrydzewski commented:
@jackspirou thanks for sharing. If this is a common situation with corporate networks, which it sounds like it might be, I'm ok with skip verify flag. Just playing devils advocate before introducing a new dependency :smile: 

my only other comment for the pull request is that we should combine wrapper and template. I understand why it was done this way, to let someone customize the body without all the boilerplate, however I think email templates are going to be too big to put in the yaml (even with this `wrapper` optimization). We should optimize for a single template and expect people to store in a gist.

tboerger commented:
I have tried both variants again, including the wrapper and excluding the wrapper for the html2text part to send mails as plaintext for a fallback. I have to say that I would prefer a separate wrapper as most people don't need to update the overall style.

Including wrapper:

```
Successful build #22 ( http://drone.mycompany.com/foo/bar/22 )

Repo: foo/bar
Author: johnsmith
Branch: master
Commit:
Time: 210ns

Update the Readme

Notification sent by Drone CI .
```

Excluding wrapper:

```
Successful build #22 ( http://drone.mycompany.com/foo/bar/22 )

Repo: foo/bar
Author: johnsmith
Branch: master
Commit:
Time: 210ns

Update the Readme
```

tboerger commented:
@bradrydzewski so what is the decision for this pull request? What do you want to keep and what do you want to have changed?

bradrydzewski commented:
my preference is to only make people update a single template (so no wrapper) and in the docs we just give them the link to a sample template in the repo they can customize. I can see most people wanting to completely start from scratch and fully brand the email template

tboerger commented:
I have dropped the wrapper part and added it as a single template. This is how the default template looks like:

<img width="820" alt="bildschirmfoto 2016-02-01 um 12 40 14" src="https://cloud.githubusercontent.com/assets/156964/12716603/08736726-c8e2-11e5-94fc-317593d33ba4.png">

Are you fine with merging it that way?

athieriot commented on DOCS.md:
```diff
@@ -22,3 +22,34 @@ notify:
+* `subject` - A handlebars template to create a custom subject. For more
+  details take a look at the [docs](http://handlebarsjs.com/).
+* `template` - A handlebars template to create a custom template. For more
+  details take a look at the [docs](http://handlebarsjs.com/).
```
You might want to add ```skip_verify``` to the documentation

tboerger commented:
The remote templates depend on https://github.com/drone/drone-go/pull/18

bradrydzewski commented:
LGTM

tboerger commented:
LGTM
